src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -346,6 +453,210 @@ private BoundExpression RewriteStringConcatenationManyExprs(SyntaxNode syntax, I
             return BoundCall.Synthesized(syntax, receiverOpt: null, initialBindingReceiverIsSubjectToCloning: ThreeState.Unknown, method, array);
         }
 
+        private bool TryRewriteStringConcatenationWithSpanBasedConcat(SyntaxNode syntax, ImmutableArray<BoundExpression> args, [NotNullWhen(true)] out BoundExpression? result)
+        {
+            // As we scan arguments we might face span-based string.Concat. If it got to this stage it must be a user-provided one (since we weren't able to unwrap it back to strings previously)
+            // We try our best to merge it with another argument to minimize amount of intermediate string allocations,
+            // e.g. turn `string.Concat(span1, span2) + string3` into `string.Concat(span1, span2, string3ConvertedToSpan)`
+            // instead of `string.Concat(string.Concat(span1, span2), string3)`. We do that by separately tracking arguments as is
+            // and arguments if we unwrap all user-defined `string.Concat`s. If at the end amount of arguments with unwrapped user-defined `string.Concat`s
+            // is < 5 and we have respective span-based concat member, we emit it, otherwise treat user-provided `string.Concat`s as ordinary arguments
```

Comment:
This sounds unnecessary complicated to me and I do not think the complexity is worth it. So, I suggest sticking to the recommended approach of passing only "trivial" operands that never expand to more than one operand.